# 灵敏度分析

### 例 5.7  
一家奶制品加工厂用牛奶生产 A、B 两种奶制品，1 桶牛奶可以在甲类设备上用 12h 加工成 3kg A，或者在乙类设备上用 8h 加工成 4kg B。假定根据市场需求，生产的 A、B 全部能售出，且每千克 A 获利 24 元，每千克 B 获利 16 元。现在加工厂每天能得到 50 桶牛奶的供应，每天正式工人总的劳动时间为 480h，并且甲类设备每天至多能加工 100kg A，乙类设备的加工能力没有限制。试为该厂制订一个生产计划，使每天获利最大，并进一步讨论以下两个附加问题：  
(1) 若可以聘用临时工人以增加劳动时间，是否聘用临时工人。  
(2) 假设由于市场需求变化，每千克 A 的获利增加到 30 元，是否改变生产计划。  


### 解  
(1) 设每天用 $ x_1 $ 桶牛奶生产 A，用 $ x_2 $ 桶牛奶生产 B，每天获利 $ z $ 元。根据题意建立问题的数学模型为  
$$
\begin{align*}
\max &\quad z = 72x_1 + 64x_2, \\
\text{s.t.} &\quad 
\begin{cases} 
x_1 + x_2 \leqslant 50, \\
12x_1 + 8x_2 \leqslant 480, \\
3x_1 \leqslant 100, \\
x_1 \geqslant 0,\ x_2 \geqslant 0.
\end{cases}
\end{align*}
$$

In [1]:
from scipy.optimize import linprog

c = [-72, -64]
A = [[1, 1], [12, 8]]
b = [[50], [480]]
x1_bounds = (0, 100/3)
x2_bounds = (0, None)
res = linprog(c, A_ub=A, b_ub=b, A_eq=None, b_eq=None, bounds=[x1_bounds, x2_bounds], method='highs', options={"disp": True})
print(f"求解结果如下:\n{res}")

求解结果如下:
        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -3360.0
              x: [ 2.000e+01  3.000e+01]
            nit: 2
          lower:  residual: [ 2.000e+01  3.000e+01]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [ 1.333e+01        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  0.000e+00]
                 marginals: [-4.800e+01 -2.000e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0


`message: Optimization terminated successfully. (HiGHS Status 7: Optimal) ` 
描述求解器运行的最终状态。表示优化问题成功求解，HiGHS状态码7表示找到最优解。

`success: True `   
布尔值，表示求解是否成功。True表示成功找到解，False表示失败（如不可行或无界）。

`status: 0 `   
求解器返回的状态码。0通常表示成功（最优解），其他值可能表示不可行（infeasible）、无界（unbounded）等，具体含义依赖求解器。  

`fun: -3360.0`  
目标函数的最优值。即优化问题的最优目标值，这里为-3360.0（负值可能表示最大化问题或目标函数定义为负）。

`x: [2.000e+01 3.000e+01]`  
最优解向量，表示决策变量的值。这里有两个变量，值为20和30。

`nit: 2`  
求解器的迭代次数。表示求解器找到最优解所需的迭代步数，这里为2次。
`lower:  `  
`residual: [2.000e+01 3.000e+01] `   
下界残差，表示变量值相对于下界的差（这里变量值为20和30，可能下界为0，残差即变量值本身）。  
`marginals: [0.000e+00 0.000e+00] `   
下界的拉格朗日乘子（对偶变量），表示下界约束的敏感性。0表示下界约束未激活（变量未被约束在下界）。

`upper: `   
`residual: [1.333e+01 inf]  `  
上界残差，表示变量值到上界的距离。1.333e+01表示第一个变量距上界13.33，inf表示第二个变量无上界或上界远大于变量值。  
`marginals: [0.000e+00 0.000e+00]`  
上界的拉格朗日乘子。0表示上界约束未激活。


`eqlin:`  
`residual: []`  
等式约束的残差（实际值与约束值的差）。空列表表示无等式约束。  
`marginals: []`  
等式约束的拉格朗日乘子。空列表表示无等式约束。


`ineqlin:`  
`residual: [0.000e+00 0.000e+00]`   
不等式约束的残差（约束值与实际值的差）。0表示不等式约束被满足（可能为等式或松弛约束）。  
`marginals: [-4.800e+01 -2.000e+00]`  
不等式约束的拉格朗日乘子，负值表示约束对目标函数的敏感性。非零值（如-48和-2）表示约束激活，改变约束会影响目标值。

`mip_node_count: 0`    
混合整数规划（MIP）的节点数。0表示问题为线性规划（无整数变量）或未使用分支定界法。  
`mip_dual_bound: 0.0`  
混合整数规划的对偶界。0表示非MIP问题或对偶界无意义。  
`mip_gap: 0.0`  
混合整数规划的相对间隙（最优解与对偶界的差距）。0表示问题为线性规划或已精确求解。

总结：结果表示优化问题成功求解，最优值为-3360，变量值为[20, 30]，迭代2次完成。约束均满足（残差为0），不等式约束激活（拉格朗日乘子非零）。问题为线性规划（无MIP相关信息）。

即最优解为 $ x_1 = 20 $，$ x_2 = 30 $，最大收益为 3360 元。  

从 slack 的两个分量都为 0 可以知道，两个约束条件都是 “紧约束”，即最优解使得不等式的约束条件达到了边界，约束条件实际上是等式约束。所以增加劳动时间，会提高收益，因而附加问题 (1) 中应该聘用临时工人。  

(2) 由于 Python 软件灵敏度分析做得不太好，附加问题 (2) 中参数变化时，只能用软件重新计算，求得的最优解不变，最优值为 3720 元。所以不应该改变生产计划。